### PR TITLE
cp: close file descriptors after cow on linux

### LINKED ISSUE #2312: cp Too many open files (os error 24)
I get the following error when copying the [vim source package](https://github.com/vim/vim/releases/tag/v8.2.2567) in nixpkgs

```
cp -pr --reflink=auto -- /nix/store/dvh1x2m7gvp1iv5jjlvrf5ry5g3ay23k-source source
cp: error: '/nix/store/dvh1x2m7gvp1iv5jjlvrf5ry5g3ay23k-source/src/INSTALL' -> 'source/src/INSTALL': Too many open files (os error 24)
```

The smalled reproduced I came up with is ``cp -r --reflink=auto -- /nix/store/dvh1x2m7gvp1iv5jjlvrf5ry5g3ay23k-source source`` with this tarball [dvh1x2m7gvp1iv5jjlvrf5ry5g3ay23k-source.zip](https://github.com/uutils/coreutils/files/6566667/dvh1x2m7gvp1iv5jjlvrf5ry5g3ay23k-source.zip).

### PR DESCRIPTION
**Pull request by miDeb**
Instead of using into_raw_fd(), which transfers ownership and
requires us to close the file descriptor manually,
use as_raw_fd(), which does not transfer ownership to us but drops the
file descriptor when the original file is dropped (in our case at the
end of the function).

Closes #2312.
I tried to come up with a test for this, but I couldn't find a way to query the amount of file descriptors used, or to limit the number of open files for a test. If somebody could come up with a way to test this, that'd be awesome.

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
Return current file descriptor limit: ``ulimit -n``
Set it: ``ulimit -n 10``
get it of a process: ``cat /proc/<PID>/limits``

**Comment by miDeb** [comment]
I have also found `ulimit`, but how can I invoke it in a test (so that it sets the limit for the current test only)?

**Comment by miDeb** [comment]
I think I have figured out a working solution (using rlimits), please check if it looks right...

**Comment by SuperSandro2000** [comment]
> I think I have figured out a working solution (using rlimits), please check if it looks right...

I don't know rust well enough to review that but I can say that it fixed the problem in NixOS when building ed with uutils-coreutils as gnu-coreutils replacement.

File: tests/common/util.rs [hunk 1]
```diff
@@ -724,6 +729,8 @@ pub struct UCommand {
     stdout: Option<Stdio>,
     stderr: Option<Stdio>,
     bytes_into_stdin: Option<Vec<u8>>,
+    #[cfg(target_os = "linux")]
+    limits: Vec<(rlimit::Resource, rlim, rlim)>,
```
Comment (sylvestre) [inline tests/common/util.rs:733]: smart (adding it as part of the test framework)

**Pull request merged**